cmd/earthly/app/run.go

```diff
@@ -259,6 +244,49 @@ func (app *EarthlyApp) run(ctx context.Context, args []string) int {
+			matches, _ := stringutil.NamedGroupMatches(msg, buildMinutesRegex)
+			if len(matches["msg"]) > 0 {
+				msg = matches["msg"][0]
+			}
+			msg = fmt.Sprintf("%s (%s)", msg, stringutil.Title(billing.Plan().Tier))
+			app.BaseCLI.Console().VerboseWarnf(err.Error())
+			app.BaseCLI.Logbus().Run().SetGenericFatalError(time.Now(), logstream.FailureType_FAILURE_TYPE_OTHER, msg)
+			switch billing.Plan().Tier {
+			case billingpb.BillingPlan_TIER_LIMITED_FREE_TIER:
+				app.BaseCLI.Console().HelpPrintf("Visit your organization settings to verify your account\nand get 6000 free build minutes per month: %s\n", billing.GetBillingURL(app.BaseCLI.CIHost(), app.BaseCLI.Flags().OrgName))
+			case billingpb.BillingPlan_TIER_FREE_TIER:
+				app.BaseCLI.Console().HelpPrintf("Visit your organization settings to upgrade your account: %s\n", billing.GetUpgradeURL(app.BaseCLI.CIHost(), app.BaseCLI.Flags().OrgName))
+			}
+			return 1
+		case grpcErrOK && grpcErr.Code() == codes.PermissionDenied && maxSatellitesRegex.MatchString(grpcErr.Message()):
+			msg := grpcErr.Message()
+			matches, _ := stringutil.NamedGroupMatches(msg, maxSatellitesRegex)
+			if len(matches["msg"]) > 0 {
+				msg = matches["msg"][0]
+			}
+			msg = fmt.Sprintf("%s %s", stringutil.Title(billing.Plan().Tier), msg)
+			app.BaseCLI.Console().VerboseWarnf(err.Error())
+			app.BaseCLI.Logbus().Run().SetGenericFatalError(time.Now(), logstream.FailureType_FAILURE_TYPE_OTHER, msg)
+			switch billing.Plan().Tier {
+			case billingpb.BillingPlan_TIER_LIMITED_FREE_TIER:
+				app.BaseCLI.Console().HelpPrintf("Visit your organization settings to verify your account\nfor an option to launch more satellites: %s\nor consider removing one of your existing satellites (`earthly sat rm <satellite-name>`)", billing.GetBillingURL(app.BaseCLI.CIHost(), app.BaseCLI.Flags().OrgName))
+			case billingpb.BillingPlan_TIER_FREE_TIER:
+				app.BaseCLI.Console().HelpPrintf("Visit your organization settings to upgrade your account for an option to launch more satellites: %s.\nAlternatively consider removing one of your existing satellites (`earthly sat rm <satellite-name>`)\nor contact support at support@earthly.dev to potentially increase your satellites' limit", billing.GetUpgradeURL(app.BaseCLI.CIHost(), app.BaseCLI.Flags().OrgName))
+			default:
+				app.BaseCLI.Console().HelpPrintf("Consider removing one of your existing satellites (`earthly sat rm <satellite-name>`)\nor contact support at support@earthly.dev to potentially increase your satellites' limit")
+			}
+			return 1
+		case grpcErrOK && grpcErr.Code() == codes.PermissionDenied && requestIDRegex.MatchString(grpcErr.Message()):
+			msg := grpcErr.Message()
+			matches, _ := stringutil.NamedGroupMatches(msg, requestIDRegex)
+			if len(matches["msg"]) > 0 {
+				msg = matches["msg"][0]
+			}
+			app.BaseCLI.Console().VerboseWarnf(err.Error())
+			app.BaseCLI.Logbus().Run().SetGenericFatalError(time.Now(), logstream.FailureType_FAILURE_TYPE_OTHER, msg)
```

Comment:
Just a thought regarding these hints. A lot of this logic seems tightly related to Earthly Cloud. I wonder if it would have been better for the server to format these different messages (and doing the switch-case's over different plans), and the client would just display whatever the server says. It would make it easier to change in the future, and it would have kept coupling low.

Just a thought for the future - no need to change anything in this PR, since this might be a bigger change.

Reply:
I agree.
for example I tried to avoid words like, "upgrade and increase to 10 satellites" because I didn't know what the plan max might be in the future.

I would actually consider decoupling further by exposing APIs that:
1. return the default values for each plan as they are currently set (e.g. how many satellites can you use in a tier plan right now?)
2. and/or APIs to return the exact text by plan. because I don't think it's compute's job (in this case) to return the too much information AND that it should be ok for compute to change its error messages without breaking how errors are displayed in earthly